### PR TITLE
SymPy CZI Wrapup Blog Post

### PR DESCRIPTION
**Pull request by moorepants**
@brocksam I've starting drafting up a wrap up blog post. Will work on it over the next month as we finalize things.

Outline and writing tasks:

- [x] Quick recap of what the grant was about and what our specific goals were.
- [x] Improvements to SymPy Mechanics
  - [x] Joints (maybe just mention/link to Timo's GSoC report)
  - [x] Zero divisions and numerical stability of the linear solves in `KanesMethod` and `LInearizer`, bicycle model fixed
  - [x] `Force`, `Torque`, `Inertia`, and `Actuator` classes
  - [x] `System`
- [x] Improvements to code gen
  - [x] lambdify speed up (doesn't print full docstring), can now evaluate the bike model with numpy (couldn't before)
  - [x] Support for passing CSEd expressions: https://github.com/sympy/sympy/pull/24649 and https://github.com/sympy/sympy/pull/25797
  - [x] Faster Jacobians in CSEd form for codegen: https://github.com/sympy/sympy/pull/25801
- [x] Introduction to SymPy Biomechanics
- [x] BRiM
- [x] Optimal Control
  - [x] PyCollo documentation and bug fixes
  - [x] Opty cse differentiation
- [x] Solving the skateboard optimal control problem
- [x] Solving muscle driven bicycle steerer
- [x] Future ideas

### CODE REVIEW

**Comment by moorepants** [comment]
@TJStienstra would you be willing to upload your literature study to a public repository for sharing? If so, I could link to it here.

File: content/czi-sympy-wrapup.rst [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+.. _SymPy: https://www.sympy.org
+
+General Improvements to SymPy Mechanics
+=======================================
+
+Joints Package
+--------------
+
+Timo Steinstra began with the project through a 2022 Google Summer of Code
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:76]: 'Stienstra'

File: content/czi-sympy-wrapup.rst [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+   Muscle driven arm pushing and pulling a lever taken from the new tutorial.
+
+.. _sympy.physics.biomechanics: https://docs.sympy.org/dev/modules/physics/biomechanics/index.html
+
+Code Generation
+===============
+
+``lambdify()`` is the primary interface for converting SymPy expressions into
+NumPy-backed Python functions for numerical evaluation. ``lamdify()`` has not
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:273]: NumPy-backend

File: content/czi-sympy-wrapup.rst [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+Framework
+<http://resolver.tudelft.nl/uuid:a2b132e9-8d38-4553-8587-0c9e3341b202>`__",
+details the design, implementation, and use of BRiM. We also wrote a paper,
+"`BRiM: A Modular Bicycle-Rider Modeling Framework
+<https://doi.org/10.59490/6504c5a765e8118fc7b106c3>`__", for the Bicycle and
+Motorcycle Dynamics 2023 conference that gives a more concise overview of the
+package as well as demonstrating easily swapping models for optimal control
+results.
+
+.. figure:: https://tjstienstra.github.io/brim/_images/lane_change.gif
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:373]: `https://tjstienstra.github.io/brim/_static/lane_change.gif` is probably more stable, I'm surprised that it actually maintained backward compatibility to that link.

File: content/czi-sympy-wrapup.rst [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+   :align: center
+   :width: 80%
+
+   Lane change simulation created with BRiM showing without and without a rider
+
+- BRiM source code: https://github.com/TJStienstra/brim/
+- BRiM documentation: https://tjstienstra.github.io/brim/
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:380]: We should probably also just directly point to the paper repo, so you can see the optimization source code: `https://github.com/TJStienstra/brim-bmd-2023-paper`

**Comment by TJStienstra** [comment]
> @TJStienstra would you be willing to upload your literature study to a public repository for sharing?

I can upload the pdf somewhere and create a link, [e.g.](https://drive.google.com/file/d/14NqIhxZcHH_oXtMMtuGeolAL2MXFGnol/view?usp=share_link).

**Comment by moorepants** [comment]
How about figshare? or zenodo? for more permanency.

**Comment by moorepants** [comment]
@brocksam and @TJStienstra this is in decent enough shape for review. I have some more things to add, but I can take comments now. If you have some time to review, that is welcomed. I want to get this published in the next couple of days so that I can then take content from this post for the CZI funding report.

File: content/czi-sympy-wrapup.rst [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+variables involved.
+
+.. _Gaussian elimination: https://en.wikipedia.org/wiki/Gaussian_elimination
+
+There are four ways, it seems, to deal with this:
+
+1. select the generalized coordinates, generalized speeds, and constants such
+   that divide-by-zero cannot occur for the numerical values of interest
+2. select symbolic Gaussian elimination algorithms that do not put the
+   solutions in a form that have divide-by-zero for the numerical values of
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:148]:
```suggestion
   solutions in a form that has divide-by-zero for the numerical values of
```

File: content/czi-sympy-wrapup.rst [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+1. choosing generalized coordinates and speeds
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+TODO : Link to the paper that Luke cites.
+
+2. Alternative Symbolic Solvers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In 2014, we switched to using ``LUsolve()`` for all of the linear solves in
+SymPy Mechanics in `PR 7581`_, which resulted in a unnoticed regression of
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:162]:
```suggestion
SymPy Mechanics in `PR 7581`_, which resulted in an unnoticed regression of
```

File: content/czi-sympy-wrapup.rst [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+divide-by-zero issues for complex problems. That change broke the crucial
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:163]:
```suggestion
divide-by-zero issues for complex problems. This change broke the crucial
```
While `this` and `that` are both possible. We are already using the word `that` in this sentence.

File: content/czi-sympy-wrapup.rst [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+We introduced three new helper classes to extend the functionality of inertia
+and loads beyond that of simply dyadics and vectors: ``Inertia()``,
+``Force()``, and ``Torque()``.
+
+The inertia object lets you associate a dyadic with a point, to completely
+define inertia for a rigid body, particle, or collection of them. Force and
+Torque are named tuples that associate a vector and point and a vector and a
+frame, respectively.
+
+We have introduced an actuator_ module that has classes that describes the
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:296]:
```suggestion
We have introduced an actuator_ module that has classes that describe the
```

File: content/czi-sympy-wrapup.rst [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+Introducing SymPy Biomechanics
+==============================
+
+We have developed a new sub-package, sympy.physics.biomechanics_, that enables
+including musculotendon force actuators in multibody dynamics models created
+with ``sympy.physics.mechanics``. ``biomechanics`` contains these primary
+modules:
+
+- ``curve.py``: contains classes that represent mathematical functional
+  relationships between the time varying muscle-tendon length, velocity, and
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:317]:
```suggestion
  relationships between the time-varying muscle-tendon length, velocity, and
```

File: content/czi-sympy-wrapup.rst [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+.. _cse(): https://docs.sympy.org/latest/modules/simplify/simplify.html#sympy.simplify.cse_main.cse
+
+Modeling and Optimal Control Uses
+=================================
+
+As explained in the introduction, our goal is to make SymPy capable of deriving
+computationally efficient neuromuscular driven multibody models. One use case
+for these models is solving `optimal control`_ problems, which benefit greatly
+from the fastest numerical evaluation of the equations of motion and its higher
+order partial derivatives. In particular, forming a `nonlinear programming`_
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:448]: `higher order` should be hyphenated: `higher-order`.

File: content/czi-sympy-wrapup.rst [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+
+- BRiM source code: https://github.com/TJStienstra/brim/
+- BRiM documentation: https://tjstienstra.github.io/brim/
+- BRiM BMD 2023 paper: https://doi.org/10.59490/6504c5a765e8118fc7b106c3
+- BRiM BMD 2023 paper source code: https://github.com/TJStienstra/brim-bmd-2023-paper
+
+Optimal Bicycle-Rider Trajectories
+----------------------------------
+
+With all of the above work, we were able to solve a optimal control problem of
```
Comment (TJStienstra) [inline content/czi-sympy-wrapup.rst:532]:
```suggestion
With all of the above work, we were able to solve an optimal control problem of
```

File: content/czi-sympy-wrapup.rst [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+.. _SymPy: https://www.sympy.org
+
+General Improvements to SymPy Mechanics
+=======================================
+
+Joints Package
+--------------
+
+Timo Stienstra began working on this project through a 2022 Google Summer of
+Code internship where he improved the SymPy Mechanics Joints package with
```
[the remark this answers is not in the document]
Comment (moorepants) [reply content/czi-sympy-wrapup.rst:80]: Any directory with an `__init__.py` is a package.

- module: .py file
- package: directory with `__init__.py`
- distribution: collection of packages (the thing released on pypi)

That's how I've always understood it.

File: content/czi-sympy-wrapup.rst [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+
+Kane's Method relies on solving three sets of linear equations:
+
+1. putting the kinematical differential equations in explicit form
+   :math:`\dot{\mathbf{q}} = \mathbf{M}_k^{-1}\left(\mathbf{u} +
+   \mathbf{f}_k\right)`
+2. putting the dynamical differential equations in explicit form
+   :math:`\dot{\mathbf{u}} = \mathbf{M}_d^{-1}\mathbf{f}_d`
+3. solving the dependent generalized speeds in terms of the independent
+   generalised speeds
```
[the remark this answers is not in the document]
Comment (moorepants) [reply content/czi-sympy-wrapup.rst:133]: Hah!

File: content/czi-sympy-wrapup.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (moorepants) [reply content/czi-sympy-wrapup.rst:80]: It's just two modules. I can adjust.

**Comment by moorepants** [comment]
Thanks Timo. 

> Maybe extend a little on the biomechanics module, as that is a rather significant contribution, where each module is now summarized in a single line.

I didn't want to repeat what we have in the tutorial, which is quite extensive. What could I add that isn't a repeat of what we have in the SymPy docs? I could add another image I suppose.

**Comment by moorepants** [comment]
Ok, this is now done. I will publish it tomorrow. @brocksam I think you are on holiday, but if you have a moment and want anything else in here or things corrected, let me know. I'll write the CZI report this weekend.

File: content/czi-sympy-wrapup.rst [hunk 14]
```diff
@@ -0,0 +1,776 @@
...
+=======================================================
+
+:date: 2023-10-27 00:00:00
+:tags: sympy, czi, biomechanics, multibody dynamics, open source software,
+       muscles, optimal control, bicycle, skateboard, code generation
+:category: research
+:slug: sympy-czi-outcomes
+:authors: Jason K. Moore
+:summary: Summary of the work performed under the Chan-Zuckerberg Foundation
+          Open Source Software for Science Cycle 4 grant awarded to Sympy, and
```
Comment (moorepants) [inline content/czi-sympy-wrapup.rst:12]:
```suggestion
          Open Source Software for Science Cycle 4 grant awarded to Sympy and
```

**Pull request merged**